File: ts/scene_io.ts
import { promises } from 'fs';

import { Material, Scene } from './proto/scene';

const readFileAsync = promises.readFile;
const writeFileAsync = promises.writeFile;

export type SceneIO = {

    decode: (data: string | Buffer) => Scene,
    read: (path: string) => Promise<Scene>,

    encode: (scene: Scene) => Uint8Array,
    save: (scene: Scene, path: string) => Promise<void>,
    encodeAsJson: (scene: Scene) => string,
    saveAsJson: (scene: Scene, path: string) => Promise<void>,
};

const preprocessScene = (scene: Scene) => {
    // workaround of what seems to be ts-proto bug.
    scene.sceneObjects
        .filter(obj => obj.material != undefined)
        .forEach(obj => (obj.material as Material).id = '');
    
    return scene;
};

const decode = (data: Buffer | string) => {
    try {
        return Scene.fromJSON(JSON.parse(data.toString()));
    } catch (e) {
        if (Buffer.isBuffer(data)) {
            return Scene.decode(data);
        } else {
            throw Error('Failed to decode as scene format: input is string, but not a valid json: ' + e);
        }
    }
};

const encode = (scene: Scene) => Scene.encode(preprocessScene(scene)).finish();
const encodeAsJson = (scene: Scene) => JSON.stringify(Scene.toJSON(scene), null, 4);

export const SceneFormatIO : SceneIO = {

    decode,
    read: async (path: string) => decode(await readFileAsync(path)),
    encode,
    save: async (scene: Scene, path: string) => await writeFileAsync(path, Buffer.from(encode(scene))),
    encodeAsJson,
    saveAsJson: async (scene: Scene, path: string) => await writeFileAsync(path, encodeAsJson(scene)),
};

File: java/sceneformat/src/main/java/com/github/computergraphicscourse/sceneformat/SceneIO.java
package com.github.computergraphicscourse.sceneformat;

import static com.github.computergraphicscourse.sceneformat.SceneFormat.*;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public interface SceneIO {

  Scene decode(byte[] data) throws IOException;
  Scene decode(String data) throws IOException;

  Scene read(InputStream inputStream) throws IOException;
  Scene read(File file) throws IOException;
  Scene read(String inputPath) throws IOException;

  byte[] encode(Scene scene) throws IOException;
  String encodeAsJson(Scene scene) throws IOException;

  void save(Scene scene, OutputStream outputStream) throws IOException;
  void save(Scene scene, File file) throws IOException;
  void save(Scene scene, String outputPath) throws IOException;

  void saveAsJson(Scene scene, OutputStream outputStream) throws IOException;
  void saveAsJson(Scene scene, File file) throws IOException;
  void saveAsJson(Scene scene, String outputPath) throws IOException;
}
